docs/programming-guide.md

```diff
@@ -1086,6 +1086,66 @@ for details.
 </tr>
 </table>
 
+### Shuffle operations
+
+Certain operations within Spark trigger an event known as the shuffle. The shuffle is Spark's
+mechanism for re-distributing data so that is grouped differently across partitions. This typically
+involves copying data across executors and machines, making the shuffle a complex and
+costly operation.
+
+#### Background
+
+To understand what happens during the shuffle we can consider the example of the
+[`reduceByKey`](#ReduceByLink) operation. The `reduceByKey` operation generates a new RDD where all
+values for a single key are combined into a tuple - the key and the result of executing a reduce
+function against all values associated with that key. The challenge is that not all values for a
+single key necessarily reside on the same partition, or even the same machine, but they must be
+co-located to compute the result.
+
+In Spark, data is generally not distributed across partitions to be in the necessary place for a
+specific operation. During computations, a single task will operate on a single partition - thus, to
+organize all the data for a single `reduceByKey` reduce task to execute, Spark needs to perform an
+all-to-all operation. It must read from all partitions to find all the values for all keys, and then
+organize those such that all values for any key lie within the same partition - this is called the
+**shuffle**.
+
+Although the set of elements in each partition of newly shuffled data will be deterministic, the
+ordering of these elements is not. If one desires predictably ordered data following shuffle
```

Comment:
Would that be affected by a shuffle? the data in the RDD are opaque; a `(key, List(value1, value2, ...))` shuffles as a unit. If you were literally shuffling that RDD.

Yeah you're talking about, say, what to do with the result of `groupByKey`, which includes a shuffle along the way, and whose ordering of keys -- and in this case of values for keys too -- was determined by the shuffle. I feel like that might be mixing two different things, though this case is also relevant. I think I'd stick to the surface level here of talking about ordering of elements in RDDs rather than values within keys.